FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/FhirToDataLakeProcessingJobSpliter.cs

```diff
@@ -0,0 +1,291 @@
+        {
+            _dataClient = EnsureArg.IsNotNull(dataClient, nameof(dataClient));
+            _diagnosticLogger = EnsureArg.IsNotNull(diagnosticLogger, nameof(diagnosticLogger));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+        }
+
+        public int LowBoundOfProcessingJobResourceCount { get; set; } = JobConfigurationConstants.LowBoundOfProcessingJobResourceCount;
+
+        public int HighBoundOfProcessingJobResourceCount { get; set; } = JobConfigurationConstants.HighBoundOfProcessingJobResourceCount;
+
+        public async IAsyncEnumerable<SubJobInfo> SplitJobAsync(string resourceType, DateTimeOffset? startTime, DateTimeOffset endTime, [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            var totalCount = await GetResourceCountAsync(resourceType, startTime, endTime, cancellationToken);
+
+            if (totalCount < HighBoundOfProcessingJobResourceCount)
+            {
+                // Return job with total count lower than high bound.
+                _logger.LogInformation($"Generate one {resourceType} job with {totalCount} count.");
+
+                yield return new SubJobInfo
+                {
+                    ResourceType = resourceType,
+                    TimeRange = new TimeRange() { DataStartTime = startTime, DataEndTime = endTime },
+                    JobSize = totalCount,
+                };
+            }
+            else
+            {
+                // Split large size job using binary search.
+                var splitingStartTime = DateTimeOffset.Now;
+                _logger.LogInformation($"Start spliting {resourceType} job, total {totalCount} resources.");
+
+                var anchorList = await InitializeAnchorListAsync(resourceType, startTime, endTime, totalCount, cancellationToken);
+
+                _logger.LogInformation($"Spliting {resourceType} job. Use {(DateTimeOffset.Now - splitingStartTime).TotalMilliseconds} milliseconds to initilize anchor list.");
+                var lastSplitTimestamp = DateTimeOffset.Now;
+                DateTimeOffset? nextJobEnd = null;
+                var jobCount = 0;
+
+                while (nextJobEnd == null || nextJobEnd < endTime)
+                {
+                    anchorList = anchorList.OrderBy(p => p.Key).ToDictionary(p => p.Key, o => o.Value);
+                    DateTimeOffset? lastEndTime = nextJobEnd ?? startTime;
+
+                    nextJobEnd = await GetNextSplitTimestamp(resourceType, lastEndTime, anchorList, cancellationToken);
+
+                    var jobSize = lastEndTime == null ? anchorList[(DateTimeOffset)nextJobEnd] : anchorList[(DateTimeOffset)nextJobEnd] - anchorList[(DateTimeOffset)lastEndTime];
+                    _logger.LogInformation($"Spliting {resourceType} job. Generated new sub job using {(DateTimeOffset.UtcNow - lastSplitTimestamp).TotalMilliseconds} milliseconds with {jobSize} resource counts. ");
```

Comment:
Here use **DateTimeOffset.UtcNow** but other logs use **DateTimeOffset.Now**?

Reply:
Thanks, fixed it.